Implement expiration date for federated shares

PVince81 opened the pull request:
Fixes issue that the UI checkbox for expiration date was not being saved for federated share.

### Todos

- [x] Bring back expiration date field (revert #25933)
- [x] Setting expiration date is saved in DB
- [x] Clearing expiration date is saved in DB
- [x] Refresh the page and open the share panel: expiration date is sameas DB
- [x] The share actually expires (tested by manually setting a past date in DB)
- [x] When expiring, should notify the remote servers of the disappearance => not needed, they find out during next access
- [x] Add to NextCloud properly saves expiration date from link ~~(can't test due to another bug)~~
- [x] Default expiration date applied to remote share
- [x] Remote group thing ??
- [x] New capability for clients to recognize this support: had to add a new one, there was a bogus old one there...
- [x] Extend unit tests
- [x] Add separate setting for expiration date of federated shares instead of reusing the one from internal shares ?

### Notes

~~I do wonder if that checkbox was added by mistake in the UI through recent refactorings.
The more I dig, the more it looks like this feature was never implemented.~~
We removed it here https://github.com/nextcloud/server/pull/25933 and will add it back in this PR.

@rullzer

PVince81 committed:
Implement expiration date for federated shares

Add expiration date field in UI.
Save expiration date when creating or updating federated share.
Read expiration date from DB in federated share provider.
Applies to both federated user and group shares.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
been lucky so far, the missing bit was reading expiration date from the DB.
after that, expiration already worked automagically

PVince81 commented:
confusing report...
<img width="679" alt="image" src="https://user-images.githubusercontent.com/277525/111292861-e9033180-8648-11eb-9778-847b965f0b19.png">

I did run cs:fix locally on related folders but nothing got fixed. I'll try a full run then.

PVince81 commented:
attempt at fixing CS on master, as a result of running `composer cs:fix`: https://github.com/nextcloud/server/pull/26143

PVince81 commented:
I've also tested remote group share and made adjustment, it works and expires properly too

PVince81 committed:
Fix expire date capability for federated shares

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
Implementation is done, here are topics to discuss before I move forward with unit tests:

1) Do we want a separate setting for the default/enforced expiration of shares ?

2) Should we delete the bogus "federation.expire_data" capability ? It was there before, stating that expiration date works for federation when it doesn't.

PVince81 commented:
- [ ] tech debt: the methods `validateExpirationDateInternal` and `validateExpirationDate` in Share20/Manager.php look exactly the same (possibly a failed merge conflict resolution)

LukasReschke commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -587,15 +587,39 @@ public function createShare(
+					$expireDate = $this->parseDate($expireDate);
+					$share->setExpirationDate($expireDate);
+				} catch (\Exception $e) {
+					throw new OCSNotFoundException($this->l->t('Invalid date, date format must be YYYY-MM-DD'));
+				}
+			}
 		} elseif ($shareType === IShare::TYPE_REMOTE_GROUP) {
 			if (!$this->shareManager->outgoingServer2ServerGroupSharesAllowed()) {
 				throw new OCSForbiddenException($this->l->t('Sharing %1$s failed because the back end does not allow shares from type %2$s', [$path->getPath(), $shareType]));
 			}
 
+			if ($shareWith === null) {
+				throw new OCSNotFoundException($this->l->t('Please specify a valid federated group id'));
+			}
+
 			$share->setSharedWith($shareWith);
 			$share->setPermissions($permissions);
+			if ($expireDate !== '') {
+				try {
+					$expireDate = $this->parseDate($expireDate);
+					$share->setExpirationDate($expireDate);
+				} catch (\Exception $e) {
+					throw new OCSNotFoundException($this->l->t('Invalid date, date format must be YYYY-MM-DD'));
```
Should we consider using `OCSBadRequestException` instead? 

```suggestion
					throw new OCSBadRequestException($this->l->t('Invalid date, date format must be YYYY-MM-DD'));
				}
			}
		} elseif ($shareType === IShare::TYPE_REMOTE_GROUP) {
			if (!$this->shareManager->outgoingServer2ServerGroupSharesAllowed()) {
				throw new OCSForbiddenException($this->l->t('Sharing %1$s failed because the back end does not allow shares from type %2$s', [$path->getPath(), $shareType]));
			}

			if ($shareWith === null) {
				throw new OCSBadRequestException($this->l->t('Please specify a valid federated group id'));
			}

			$share->setSharedWith($shareWith);
			$share->setPermissions($permissions);
			if ($expireDate !== '') {
				try {
					$expireDate = $this->parseDate($expireDate);
					$share->setExpirationDate($expireDate);
				} catch (\Exception $e) {
					throw new OCSBadRequestException($this->l->t('Invalid date, date format must be YYYY-MM-DD'));
```

PVince81 replied:
I didn't want to change the existing semantics as it might affect clients and what status code they are reacting on.

cc @tobiasKaminsky

PVince81 commented:
>     * tech debt the methods `validateExpirationDateInternal` and `validateExpirationDate` in Share20/Manager.php look exactly the same (possibly a failed merge conflict resolution)

WRONG, I diffed them locally and they are different. `validateExpirationDate` is for links and in one place it is used by user shares (and I copy-pasted the wrong thing in this PR). I've raised https://github.com/nextcloud/server/issues/26197 to fix this inconsistency first.

PVince81 commented:
- [ ] rebase after https://github.com/nextcloud/server/pull/26200 and also use the correct method for expiration validation

PVince81 commented:
rebased and integrated https://github.com/nextcloud/server/pull/26200

icewind1991 commented on lib/private/Share20/Manager.php:
```diff
@@ -779,6 +782,9 @@ public function createShare(IShare $share) {
 						\OCP\Security\ISecureRandom::CHAR_HUMAN_READABLE
 					)
 				);
+			} elseif ($share->getShareType() === IShare::TYPE_REMOTE || $share->getShareType() === IShare::TYPE_REMOTE_GROUP) {
```
type remote is already checked above?

PVince81 commented:
rebased due to another conflict and fixed the duplicate conditions

- [ ] retest
- [ ] clarify if we want a separate setting for the federation expiration

PVince81 commented:
a separate setting for federation expiration might make sense.
I could imagine scenarios where one only wants to set/enforce expiration for federated shares but not local ones, or use different durations for those

PVince81 committed:
Unit tests for federated share expiration date

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Separate settings for remote share expiration

Added separate settings for default and enforced expiration date for
remote shares.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I've decided to go ahead and implement the separate setting. Now pushed in https://github.com/nextcloud/server/pull/25320/commits/6a43ad25943ac61d08a258f8a155b70002e4c207

It was mostly a matter of finding where internal expiration date is used and copying/reusing for remote shares.

Tested by having internal expiration enforced to 3 days while remote expiration was enforced to 6 days. Checked that it's possible to set remote share expiration to 5 days and get no error.

PVince81 commented:
```
/drone/src/tests/acceptance/features/app-files-sharing.feature:23
```

PVince81 commented:
rebased for CI

skjnldsv commented:
failure unrelated

PVince81 commented:
rebased and recompiled

PVince81 commented:
the random failure was supposed to be fixed but is still happening: https://github.com/nextcloud/server/issues/26314

rullzer reviewed (approved):
Didn't fully test but looks good

PVince81 commented:
/compile amend /

PVince81 commented:
manual rebase + recompile
